packages/protocol/contracts/liquidity/GrandaMento.sol

```diff
@@ -0,0 +1,477 @@
+    // for extra certainty that cancelling an exchange proposal can never fail
+    // if for some reason the CELO balance of this contract is less than the
+    // recorded sell amount.
+    uint256 totalBalance = sellToken.balanceOf(address(this));
+    if (totalBalance < sellAmount) {
+      sellAmount = totalBalance;
+    }
+    return (sellToken, sellAmount);
+  }
+
+  /**
+   * @notice Using the oracle price, charges the spread and calculates the amount of
+   * the asset being bought.
+   * @dev Stable token value amounts are used for the sellAmount, not unit amounts.
+   * Assumes both CELO and the stable token have 18 decimals.
+   * @param stableToken The stableToken involved in the exchange.
+   * @param sellAmount The amount of the sell token being sold.
+   * @param sellCelo Whether CELO is being sold.
+   * @return The amount of the asset being bought.
+   */
+  function getBuyAmount(address stableToken, uint256 sellAmount, bool sellCelo)
+    public
+    view
+    returns (uint256)
+  {
+    // Gets the price of CELO quoted in stableToken.
+    FixidityLib.Fraction memory exchangeRate = getOracleExchangeRate(stableToken);
+    // If stableToken is being sold, instead use the price of stableToken
+    // quoted in CELO.
+    if (!sellCelo) {
+      exchangeRate = exchangeRate.reciprocal();
+    }
+    // The sell amount taking the spread into account, ie:
+    // (1 - spread) * sellAmount
+    FixidityLib.Fraction memory adjustedSellAmount = FixidityLib.fixed1().subtract(spread).multiply(
+      FixidityLib.newFixed(sellAmount)
+    );
+    // Calculate the buy amount:
+    // exchangeRate * adjustedSellAmount
+    return exchangeRate.multiply(adjustedSellAmount).fromFixed();
+  }
+
+  /**
+   * @notice Gets the proposal identifiers of all exchange proposals in the
+   * Proposed or Approved state.
+   * @return An array of active exchange proposals IDs.
+   */
+  function getActiveProposalIds() external view returns (uint256[] memory) {
+    return activeProposalIds.getKeys();
+  }
+
+  /**
+   * @notice Gets the oracle CELO price quoted in the stable token.
+   * @dev Reverts if there is not a rate for the provided stable token.
+   * @param stableToken The stable token to get the oracle price for.
+   * @return The oracle CELO price quoted in the stable token.
+   */
+  function getOracleExchangeRate(address stableToken)
+    private
+    view
+    returns (FixidityLib.Fraction memory)
+  {
+    uint256 rateNumerator;
+    uint256 rateDenominator;
+    (rateNumerator, rateDenominator) = getSortedOracles().medianRate(stableToken);
+    // When rateDenominator is 0, it means there are no rates known to SortedOracles.
+    require(rateDenominator > 0, "No oracle rates present for token");
+    return FixidityLib.wrap(rateNumerator).divide(FixidityLib.wrap(rateDenominator));
+  }
+
+  /**
+   * @notice Sets the approver.
+   * @dev Sender must be owner. New approver is allowed to be address(0).
```

Comment:
Is this equivalent to freezing GrandaMento until a new one is set?

Reply:
It'd still allow proposals in the `Approved` state to be executed once the veto period is up.

Though I could see an argument for making the contract more explicitly "frozen" when the approver == address(0). If there's something badly wrong with the multisig, likely the first & easiest reaction would be a governance proposal or hotfix to set the approver to address(0), and to then take more time to figure out the details of a new approver multisig. We could possibly add a check in `executeExchangeProposal`: `require(approver != address(0))`, and then in `cancelExchangeProposal` we allow anyone to cancel an active exchange proposal (ie refund the deposit) if `approver == address(0)`. This would probably only be useful if the multisig was compromised and was approving exchanges it shouldn't be.

Another option is to have that freezing functionality be determined more explicitly by `Freezable` or by a `frozen` boolean. Any thoughts?